lightning/src/routing/scoring.rs

```diff
@@ -666,19 +669,22 @@ impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiqui
 	/// Returns a penalty for routing the given HTLC `amount_msat` through the channel in this
 	/// direction.
 	fn penalty_msat(&self, amount_msat: u64, liquidity_penalty_multiplier_msat: u64) -> u64 {
-		let max_penalty_msat = liquidity_penalty_multiplier_msat.saturating_mul(2);
 		let max_liquidity_msat = self.max_liquidity_msat();
 		let min_liquidity_msat = core::cmp::min(self.min_liquidity_msat(), max_liquidity_msat);
-		if amount_msat > max_liquidity_msat {
-			max_penalty_msat
-		} else if amount_msat <= min_liquidity_msat {
+		if amount_msat <= min_liquidity_msat {
 			0
+		} else if amount_msat > max_liquidity_msat {
+			u64::max_value()
+		} else if amount_msat == max_liquidity_msat && max_liquidity_msat != self.capacity_msat {
```

Comment:
I mean, sure, but is it really worth constraining this vs just rounding off-by-one in the previous if?

Reply:
There are a couple edge cases to deal with.

The first mentioned in the comment is to allow at least one attempt with a non-`u64::max_value` penalty when attempting to pay with the entire effective capacity, which may be the max HTLC. We currently prefer that over capacity if both are known but should be doing the opposite. I have a commit to fix that, which we dropped from the original `ProbabilisticScorer` PR. But a user may also not have the capacity if they don't have access to the UTXO set via `chain::Access`.

The second is for private hops, where we use `u64::max_value` for the effective capacity. We'd have to use `saturated_add` so seems like a branch would be necessary. Or if we used `amount_msat  >= max_liquidity_msat` instead of adding, we'd still need to special case the first edge case, which gets a bit more messy... I suppose something like this may work since we can use `max_penalty_msat` to avoid weird logic to hit the log case?

```rust
if amount_msat >= max_liquidity_msat {
    if amount_msat > max_liquidity_msat {
        u64::max_value()
    } else if max_liquidity_msat != self.capacity_msat {
        u64::max_value()
    } else {
        max_penalty_msat
    }
}
```

That way the additional branching would be limited. How does that sound?